### PR TITLE
feat: support try-runtime for all our runtimes

### PR DESCRIPTION
**Pull request by aliXsed**
Ready for your review after passing the following tests:

- [x] Test on-runtime-upgrade sub command with parachain
- [x] Test execute-block sub command with parachain
- [x] Test on-runtime-upgrade with mainnet
- [x] Test try-runtime as a CI job on each PR

### CODE REVIEW

File: runtimes/staking/src/lib.rs
Comment (noisefloors) [inline runtimes/staking/src/lib.rs:401]: Great explanation. 💯

Comment (aliXsed) [reply runtimes/staking/src/lib.rs:401]: The credit goes to substrate :)

File: node/src/cli.rs
Comment (ETeissonniere) [inline node/src/cli.rs:96]: Why the two commands?

File: node/src/command.rs
Comment (ETeissonniere) [inline node/src/command.rs:479]: `Subcommand::TryRuntime(_)` could do the trick

**Review by ETeissonniere — COMMENTED** [review]
Great work!

**Comment by ETeissonniere** [comment]
@aliXsed you could add it to the GitHub actions when done, that way we always test for the migrations

File: node/src/cli.rs
Comment (aliXsed) [reply node/src/cli.rs:96]: This allows the command try-runtime to exist even when the feature is not enabled. In that case the command will just let the user know that they have to compile the node with the try-runtime feature enabled.

Comment (ETeissonniere) [reply node/src/cli.rs:96]: Have you tried https://github.com/NodleCode/chain/pull/444#discussion_r767384666?

Comment (aliXsed) [reply node/src/cli.rs:96]: I tested it now. This is what happens when the node is not built with the "try-runtime" and you try to call the "try-runtime" command:

> cargo run --release try-runtime                                               
    Finished release [optimized] target(s) in 0.59s
     Running `target/release/nodle-chain try-runtime`
Error: Input("TryRuntime wasn't enabled when building the node. You can enable it with `--features try-runtime`.")

Which seems what we should expect.

File: node/src/command.rs
Comment (aliXsed) [reply node/src/command.rs:479]: I think the aim of the parity guys had been to have as less of the try-runtime specific types and code to be exposed for the case the feature is not enabled. If I go with `Subcommand::TryRuntime(_)`, then I have to make the type try_runtime_cli::TryRuntimeCmd which is the type what is passed to `Subcommand::TryRuntime` also available for non try-runtime builds.

**Comment by aliXsed** [comment]
> @aliXsed you could add it to the GitHub actions when done, that way we always test for the migrations

@ETeissonniere This [commit](https://github.com/NodleCode/chain/pull/444/commits/7d89614f1b094f844f615933e7ee1dd218178570) should do the job for our main net. For our parachain we need to wait ro have a stable all time running collator to try the runtime upgrade on top of its state.

File: node/src/command.rs
Comment (noisefloors) [inline node/src/command.rs:449]: If we only want to build this branch on environments that are not under `test`: `#[cfg(all(not(test), feature = "try-runtime"))]. This might help with reducing overall build time per pull request.

Comment (aliXsed) [reply node/src/command.rs:449]: As discussed on Slack, it would be a bit tricky to generalize it.

**Comment by noisefloors** [comment]
The workflow related to testing is still required in order to merge according to the GitHub UI; that resolves my concerns from earlier.

**Comment by aliXsed** [comment]
> The workflow related to testing is still required in order to merge according to the GitHub UI; that resolves my concerns from earlier.

True and it will be done on every pull_request and before merging anything to master.

**Review by noisefloors — APPROVED** [review]

File: node/src/cli.rs
Comment (aliXsed) [reply node/src/cli.rs]: Done https://github.com/NodleCode/chain/pull/444/commits/34181a8a2c0ed4211f03e40a357db39ac1383f64

**Review by noisefloors — APPROVED** [review]
A minor change dismissed my earlier approval. It still looks good to me!

**Review by ETeissonniere — APPROVED** [review]

**Pull request merged**
